cargo-fix/tests/all/main.rs

```diff
@@ -0,0 +1,264 @@
+        };
+        let dur = start.elapsed();
+        println!("dur: {}.{:03}ms", dur.as_secs(), dur.subsec_nanos() / 1_000_000);
+        println!("exit: {}", output.status);
+        if output.stdout.len() > 0 {
+            println!("stdout ---\n{}", String::from_utf8_lossy(&output.stdout));
+        }
+        if output.stderr.len() > 0 {
+            println!("stderr ---\n{}", String::from_utf8_lossy(&output.stderr));
+        }
+        println!("···················································");
+        let code = match output.status.code() {
+            Some(code) => code,
+            None => panic!("super extra failure: {}", output.status),
+        };
+        if code != self.status {
+            panic!("expected exit code `{}` got `{}`", self.status, code);
+        }
+        self.match_std(&output.stdout, &self.stdout, &self.stdout_contains);
+        self.match_std(&output.stderr, &self.stderr, &self.stderr_contains);
+    }
+
+    fn match_std(&self, actual: &[u8], expected: &Option<String>, contains: &[String]) {
+        let actual = match str::from_utf8(actual) {
+            Ok(s) => s,
+            Err(_) => panic!("std wasn't utf8"),
+        };
+        let actual = self.clean(actual);
+        if let Some(ref expected) = *expected {
+            diff(&self.clean(expected), &actual);
+        }
+        for s in contains {
+            let s = self.clean(s);
+            if actual.contains(&s) {
+                continue
+            }
+            println!("\nfailed to find contents within output stream\n\
+                      expected to find\n  {}\n\nwithin:\n\n{}\n\n",
+                     s,
+                     actual);
+            panic!("test failed");
+        }
+    }
+
+    fn clean(&self, s: &str) -> String {
+        let url = Url::from_file_path(&self.project.root).unwrap();
+        let s = s.replace("[CHECKING]", "    Checking")
+            .replace("[FINISHED]", "    Finished")
+            .replace("[COMPILING]", "   Compiling")
+            .replace(&url.to_string(), "CWD")
+            .replace(&self.project.root.display().to_string(), "CWD")
+            .replace("\\", "/");
+        let lines = s.lines()
+            .map(|s| {
+                let i = match s.find("target(s) in") {
+                    Some(i) => i,
+                    None => return s.to_string(),
+                };
+                if s.trim().starts_with("Finished") {
+                    s[..i].to_string()
+                } else {
+                    s.to_string()
+                }
+            });
+        let mut ret = String::new();
+        for (i, line) in lines.enumerate() {
+            if i != 0 {
+                ret.push_str("\n");
+            }
+            ret.push_str(&line);
+        }
+        ret
+    }
+}
+
+impl<'a> Drop for ExpectCmd<'a> {
+    fn drop(&mut self) {
+        if !self.ran {
+            panic!("forgot to run this command");
```

Comment:
Heh, interest way to write `#[must_use]` and session types ;)

Reply:
True! This has worked out pretty well in Cargo as it's not absolutely crucial to nail ergonomcis but rather just catch accidental mistakes here and there